torch/csrc/autograd/engine.cpp

```diff
@@ -72,23 +73,38 @@ struct GraphTask {
   // Notified when a task finishes executing.  Check outstanding_tasks to see
   // if all tasks are done.
   std::condition_variable not_done;
-  const Engine::pre_callback_map& pre_callbacks;
-  const Engine::post_callback_map& post_callbacks;
   std::unordered_map<Function*, InputBuffer> not_ready;
   std::unordered_map<Function*, int> dependencies;
 
+  struct ExecInfo {
+    struct Capture {
+      Capture(int input_idx, int output_idx) : input_idx(input_idx), output_idx(output_idx) {}
+      int input_idx; // within Function inputs
+      int output_idx; // within the output vector of a GraphTask
+    };
+
+    bool needed;
+    std::unique_ptr<std::vector<Capture>> captures;
```

Comment:
It's really only one state. If we ever allocate a vector, it's only because we will fill it right after that. I didn't want to make it a vector, because a vast majority of nodes don't need this entry and I wanted to limit the number of allocations we have to do.

Reply:
I believe an empty std::vector just takes up just three words inline in libstdc++. (No extra malloc/new until you add elements).

Reply:
yeah, that's still 3x what a pointer is

Reply:
That's not a good enough reason to wrap it in a unique_ptr.

Reply:
Considering that 99% of these structs won't use the vector at all?